Add zsh terminal prompt support

nilason opened the pull request:
This addresses https://github.com/OSGeo/grass/issues/719, regarding missing GRASS specific terminal prompt with zsh.

The code is more-or-less a copy of `bash_startup()` with `bash` replaced with `zsh` at appropriate places and adopted `PS1` to zsh.

This seems to work as is, but `zsh_startup()` might be merged with `bash_startup()` using variables and I'm not familiar with every part of the code (and how to test it) – I therefore leave this as WIP.

For testing this PR : do `chsh -s $(which zsh)` and open new terminal.

nilason commented:
I have tested every bit I could think of: reading from `~/.grass.zshrc` and `~/.grass7/zshrc`, copying aliases from `~/.alias`, running `grass_prompt` and with this update of this PR I believe this is good to go.

I have seen there has been discussions on reading alias [trac 4007](https://trac.osgeo.org/grass/ticket/4007) and in comments on #170 –– I took the liberty to "hardcode" the path to `~/.alias`. I realise this may not be a good solution for a multiuser environment (with which I have no experience of with GRASS), but there are other hardcoded settings in `$LOCATION/.zshrc` such as HOME and PATH...

nilason commented:
Come to think of it: the `$LOCATION/.zshrc` is (re-)written at every startup, so the multiuser scenario mentioned above isn't an issue. I cannot think of any reason not to expand the path to `~/.alias`.

nilason commented:
Latest update will produce a result like:

```
grass79
Starting GRASS GIS...
...
Mapset <PERMANENT> in Location <nc_spm_08_grass7>                               
GRASS 7.9.dev : ~ > r.mask raster=aspect
All subsequent raster operations will be limited to the MASK area. Removing
or renaming raster map named 'MASK' will restore raster operations to
normal.
Mapset <PERMANENT> in Location <nc_spm_08_grass7>                               
GRASS 7.9.dev : ~ > g.mapset user1                        [Raster MASK present]
Mapset switched. Your shell continues to use the history for the old mapset
Mapset <user1> in Location <nc_spm_08_grass7>                                   
GRASS 7.9.dev : ~ > g.mapset PERMANENT
Mapset switched. Your shell continues to use the history for the old mapset
Mapset <PERMANENT> in Location <nc_spm_08_grass7>                               
GRASS 7.9.dev : ~ > r.mask -r                             [Raster MASK present]
Raster MASK removed
Mapset <PERMANENT> in Location <nc_spm_08_grass7>                               
GRASS 7.9.dev : ~ > 
```

I can't help myself fiddling with this, would appreciate if someone took a look at this and put an end to it :smile:
Jokes aside, zsh is with macOS 10.15 the default shell so this is a needed improvement.

HuidaeCho commented:
> About the similarities with Bash, I'm not comfortable with the large overlap, but I don't have any good solution to account for the tiny differences all over the place.

Just curious if there are major differences between bash and zsh to require separate but substantially overlapping code. Looking at `zsh_startup()`, I feel more than 50% of the code is copy-paste from `bash_startup()`. I would recommend to add a flag to `bash_startup()` and reuse the code instead of duplicating it. It will be much easier to maintain the code later.

> 
> As for the behavior with rc files and aliases, sourcing sounds good, but I can't quite get the whole picture at this point reading again discussion in #170. Perhaps @HuidaeCho has a better idea.
> 

Currently, all this startup code is a little bit messy. For example, like @wenzeslaus mentioned in [#2681](https://trac.osgeo.org/grass/ticket/2681), `LOCATION` vs. `LOCATION_NAME`. `LOCATION` is actually NOT the name of a location. It's very misleading.

For bash, I didn't source `~/.bashrc` from the startup code and left it to the user. By simply adding `source ~/.bashrc` in `~/.grass7/bashrc`, you can source this file if you want. If you want a cleaner environment, don't source it. That's a user choice.

> But still, looking on the quotations in #170, have you tried not doing the HOME change hack? The statement there by Glynn Clements ("AFAICT, the reason for messing about with the shell startup was to force the use of per-mapset history files") suggests that if history is handled through other means (here HISTFILE), the HOME hack is not needed, doesn't it?

Exactly! It would be ideal to remove the HOME hack altogether. Why not just set `HISTFILE` to `$MAPSET_PATH\.bash_history`? We need to seriously revisit the startup code.

HuidaeCho commented on lib/init/grass.py:
```diff
@@ -1901,6 +1901,94 @@ def csh_startup(location, location_name, mapset, grass_env_file):
     return process
 
 
+def zsh_startup(location, location_name, grass_env_file):
```
I would recommend to add a new boolean flag `zsh` (`False` by default) to `bash_startup()` instead of duplicating most of the code from the latter function. It will help us maintain the code later.

nilason replied:
I choose to rename to `shell_startup(location, location_name, grass_env_file, shell_name)` and pass to `shell_name` either bash or zsh.
Semantically, it felt awkward with a zsh flag to a bash function.

nilason commented:
Thanks @wenzeslaus and @HuidaeCho for taking time and effort to look into this.

As I mentioned in initial comment an option would be to merge the two functions into one as you both also mentioned. Now I did that in last push. Contrary to my fear, I believe this didn't lead to a too complicated code.
I strived to leave bash as it was. I only did source for zsh .alias, but this would be possible for bash as well (the present line: ` f.write("test -r ~/.alias && . ~/.alias\n")` doesn't do anything as $HOME at this stage is in $LOCATION and there is no .alias file there).
I hesitated to remove $LOCATION variable, maybe something for GRASS 8? But if you think now is the time, I'll remove it.

Regarding avoiding the HOME hack. I have not yet tried it, but I'm confident it is feasible. Yet perhaps not a trivial one, better left for GRASS 8 I believe. My hope for this PR is to backport it to 7.8 as well (one reason I initially added a separate method for zsh). Apart from setting HISTFILE, measures has to done to contrive support for the various shells (for setting prompts etc.).

I changed the prompt for zsh to look a bit more like bash:
```
[Raster MASK present]  
GRASS 7.9.dev : ~ >                    Mapset <PERMANENT> in <nc_spm_08_grass7>
```
the mask info (the PROMPT_COMMAND) is only printed if set, as now is the case for bash. But I lifted the location from PS1 and added this to RPROMPT together with mapset. The reason for this is that the prompt can be very long with a long location name and this in particular when working directory is not home, and I miss the mapset.

I sneaked in a fix for [trac issue 3009](https://trac.osgeo.org/grass/ticket/3009) (failure to quit GRASS from GUI on macOS and FreeBSD) adding `f.write("trap \"exit\" TERM\n")
`. If you think this may cause problems on some platforms this has to be put in a `if clause`.

nilason commented on lib/init/grass.py:
```diff
@@ -1901,6 +1901,94 @@ def csh_startup(location, location_name, mapset, grass_env_file):
+    # setting LOCATION for backwards compatibility
+    f.write("""grass_prompt() {{
+    MAPSET_PATH="`g.gisenv get=GISDBASE,LOCATION_NAME,MAPSET separator='/'`"
+    LOCATION="$MAPSET_PATH"
```
After latest push, bash and zsh are processed in one and the same function. Should I remove LOCATION for bash as well, or leave it for GRASS 8?

nilason replied:
I'll leave LOCATION for now, we'll come back for it for 8.

nilason commented:
> To make zsh fully supported, you can also add it to g.mapset to provide a proper "fix history message":

Thanks for the reminder. Added.

nilason commented:
@lbartoletti It would be great if you could test this on FreeBSD, if the zsh prompt works at all, but also if the GUI "Quit GRASS" (see #408 for previous discussion) works (with bash or zsh).
(for testing this there's no need to recompile, only copy `grass.py`)

In case a condition will be needed, what result gives python command `import sys; sys.platform` on FreeBSD? I assume a check: `isFreeBSD = 'freebsd' in sys.platform` would be good.

wenzeslaus commented on lib/init/grass.py:
```diff
@@ -1900,10 +1900,21 @@ def csh_startup(location, location_name, mapset, grass_env_file):
     os.environ['HOME'] = userhome
     return process
 
+def shell_startup(location, location_name, grass_env_file, shell_name):
```
In light of #768, `shell` should probably mean any shell in grass.py. Could this be `bourne_like_startup` or `sh_like_startup` or `sh_startup`?

HuidaeCho replied:
I agree with @wenzeslaus. `shell_startup` is too inclusive when this function only handles `bash` and `zsh`. There is `csh_startup` for another shell already, so again, semantically, `shell_startup` is not good either. If you felt awkward passing a `zsh` flag to `bash_startup` (which I'm OK with), maybe, simply `sh_startup` would be better.

However, `/bin/sh` (true `sh` shell) actually uses `default_startup`, `csh` and `tcsh` share `csh_startup`, `bash` and `msh` share `bash_startup`. Admittedly, these shells don't pass a shell-specific flag like `zsh`.

What about `bash_like_startup` or `bash_zsh_startup` (`bazsh_startup`.. let's be more creative)? It may be semantically more correct. There is no perfect solution when you combine code for multiple shells.

nilason replied:
> There is no perfect solution when you combine code for multiple shells.

Soo, true! Considering all the alternatives you mentioned and what I in addition could think of, may I suggest we choose `sh_startup`?
(Although bazsh_startup was not bad either, albeit a bit cryptic 😄)

HuidaeCho replied:
> Soo, true! Considering all the alternatives you mentioned and what I in addition could think of, may I suggest we choose `sh_startup`?
> (Although bazsh_startup was not bad either, albeit a bit cryptic smile)

`bazsh_startup` was a joke ;-). My only problem with `sh_startup` is that `sh` doesn't use `sh_startup`. I believe, like @wenzeslaus mentioned, `sh_like_startup` would be better in this case.

HuidaeCho replied:
I would rename `shell_name` to `sh` to be consistent with the rest of the script. `shellname` is used for the full name of a shell outside and `sh` for a short name. `shell_name` is the same as `sh`.

nilason replied:
> My only problem with sh_startup is that sh doesn't use sh_startup. I believe, like @wenzeslaus mentioned, sh_like_startup would be better in this case.

I can live with `sh_like_startup`.

Sidenote: I did notice that Bourne shell (sh) is using `default startup` and does not have a "fake" home in $LOCATION (hence no separate history for each location). Is this intended design or has there been no demand for this?

HuidaeCho replied:
I guess it's the latter. No demand.

HuidaeCho commented on lib/init/grass.py:
```diff
@@ -1900,10 +1900,21 @@ def csh_startup(location, location_name, mapset, grass_env_file):
+        sh_history = ".bash_history"
+        shrc = ".bashrc"
+        grass_shrc = ".grass.bashrc"
 
```
May I suggest to keep the consistent order of `shell_name` here as the first `if shell_name` above? `bash` first, `zsh` next.

HuidaeCho commented on lib/init/grass.py:
```diff
@@ -1915,52 +1926,77 @@ def bash_startup(location, location_name, grass_env_file):
+    
+    if shell_name == 'zsh':
+        f.write('test -r {home}/.alias && source {home}/.alias\n'.format(
+            home=userhome))
```
Does it mean `zsh` doesn't support `test -r ~/.alias && . ~/.alias`? Interesting.

nilason replied:
As I understand this, this line is written to a `.*rc` file in the new $HOME at $LOCATION. `~` is thus expanded to (the new) $HOME where there is no .alias file. This is the case, at least for mac, with both bash and zsh.

nilason replied:
To be concise:
`test -r {home}/.alias && . {home}/.alias` does also work. `source` not necessary (a period `.` works too), but expanded (original user) home is needed.

HuidaeCho replied:
> As I understand this, this line is written to a `.*rc` file in the new $HOME at $LOCATION. `~` is thus expanded to (the new) $HOME where there is no .alias file. This is the case, at least for mac, with both bash and zsh.

I see. You decided to source `/home/USER/.alias` from the startup script. For bash, it's up to the user and that's why I didn't do `source /home/USER/.alias` here. The user can always include `source ~/.alias` in their `~/.grass7/bashrc` if they want. I gave them a choice between a GRASS-specific bash environment vs. a full regular bash environment. That's a design decision.

nilason replied:
> I see. You decided to source /home/USER/.alias from the startup script. For bash, it's up to the user and that's why I didn't do source /home/USER/.alias here. The user can always include source ~/.alias in their ~/.grass7/bashrc if they want. I gave them a choice between a GRASS-specific bash environment vs. a full regular bash environment. That's a design decision.

I understand .alias can be sourced in ~/.grass7/bashrc, with which I have no fundamental problem with. But as of now that line https://github.com/OSGeo/grass/blob/c3a3f57aaff7acdc08fcbb82ed6647c91f0b7ce2/lib/init/grass.py#L1922 doesn't do anything and may be removed completely or alternatively rewrite it as I did now for zsh.

HuidaeCho replied:
> > I see. You decided to source /home/USER/.alias from the startup script. For bash, it's up to the user and that's why I didn't do source /home/USER/.alias here. The user can always include source ~/.alias in their ~/.grass7/bashrc if they want. I gave them a choice between a GRASS-specific bash environment vs. a full regular bash environment. That's a design decision.
> 
> I understand .alias can be sourced in ~/.grass7/bashrc, with which I have no fundamental problem with. But as of now that line
> 
> https://github.com/OSGeo/grass/blob/c3a3f57aaff7acdc08fcbb82ed6647c91f0b7ce2/lib/init/grass.py#L1922
> doesn't do anything and may be removed completely or alternatively rewrite it as I did now for zsh.

I don't know about the history of that line, but creating `$LOCATION/.alias` works fine.

nilason replied:
> I don't know about the history of that line, but creating $LOCATION/.alias works fine.

True, if there is a file there it will work, but by default there is none and it has to be filled with content. I'm wondering, is it a common practice for (some) pro-users to create specific .alias file for a single location?
Sourcing from /Users/theuser/.grass7/bashrc makes sense to me, but creating location specific .alias files not so much. But I may of course overlook something.

HuidaeCho replied:
I understand. Probably, it was an oversight by the original author and he intended to source `/home/USER/.alias`. Or it might have been that way until the `$HOME` hack broke it? Who knows... ;-)

nilason replied:
I could trace that line to at least [GRASS 5.0](https://github.com/OSGeo/grass-legacy/blob/e3cbac3817101039d6d98a69d3a09dd944c4fce4/src/general/init/init.sh#L564).

As I understand that code the $HOME hack was in place already, I'm curious whether the .alias line worked back then.

nilason replied:
Bash stays as is, while I choose to source alias for zsh.

HuidaeCho commented on lib/init/grass.py:
```diff
@@ -2404,10 +2442,16 @@ def main():
+            shell_process = shell_startup(mapset_settings.full_mapset,
+                                          mapset_settings.location,
+                                          grass_env_file,
+                                          "bash")
```
May I suggest to combine these two `elif`s:
```bash
else if sh in ['bash', 'zsh', 'msh', 'cygwin']:
   shell_process = ... grass_env_file, sh)
```
and handle `zsh` and the other `sh`'s inside the startup function.

nilason replied:
Sounds good to me.

nilason replied:
Per @wenzeslaus suggestion, let us leave this for another PR.
I have a feeling this might be totally reworked anyway for GRASS 8 if we attempt to remove the HOME hack.

HuidaeCho commented on lib/init/grass.py:
```diff
@@ -1900,10 +1900,21 @@ def csh_startup(location, location_name, mapset, grass_env_file):
     return process
 
+def shell_startup(location, location_name, grass_env_file, shell_name):
+    if shell_name not in ['bash', 'zsh']:
```
This case would be a programmer's mistake. Since this function supports `msh` and `cygwin` as well (maybe more in the future) and the user cannot make this error, I wouldn't check for `shell_name` here.

nilason replied:
> This case would be a programmer's mistake.

This was exactly my point, with the function naming problem this was a way to avoid potential errors. But if we go with the solution of moving the handling of the different shells, this is indeed superfluous.

nilason replied:
I reorganised the initial shell-check, but I think it doesn't hurt in this case to throw an exception if a none-supported shell name is called.

wenzeslaus commented:
I suggest to move ahead with this. I agree that the alias. history, and fake home situation should be resolved and we made advancements in our understanding, but let's resolve that in another PR. Same for wrapping all shells functions into one (or whatever is what you are heading towards). It seems to me that original intention and necessary code refactoring are done.

This conflicts with #768 which I would like to merge, but it already conflicts with master branch, so I though it would be easier to merge this first and then resolve all conflicts in #768. Alternatively, I can merge #768 and you (@nilason) resolve the conflict here (should be mostly indentation).

nilason commented:
I updated following suggestions from you both, as I deemed correct. Please do (re-)raise any remaining issue you might have.

Also, please do address the previously mentioned issue:

> I sneaked in a fix for trac issue 3009 (failure to quit GRASS from GUI on macOS and FreeBSD) adding f.write("trap \"exit\" TERM\n") . If you think this may cause problems on some platforms this has to be put in a if clause.

This could be enclosed in something like:
```
if sys.platform in in ['darwin', 'freebsd']:
```

wenzeslaus commented:
Thanks @nilason . Bash and zsh work for me. The GUI exit works for me on Linux. History works only with bash, not with zsh. However, this is an improvement as it is, so I merged and will leave it on you or others to open new issues or PRs.

nilason commented:
@HuidaeCho and @wenzeslaus Thanks!

> History works only with bash, not with zsh.

Thats the problem with not being able to test on different platforms. For a quick check @wenzeslaus, could you please run in a zsh terminal `echo "HISTSIZE <$HISTSIZE> SAVEHIST <$SAVEHIST>, HISTFILE <$HISTFILE>"` before and while running grass and compare the result with description on [A User's Guide to the Z-Shell – 2.5.4: Setting up history](http://zsh.sourceforge.net/Guide/zshguide02.html#l17).

wenzeslaus commented:
`SAVEHIST` is 0. Changing it to non-zero fixes the issue for me. Outside of the GRASS session, in zsh, it is 1000.

```
GRASS 7.9.dev : grass > echo "HISTSIZE <$HISTSIZE> SAVEHIST <$SAVEHIST>, HISTFILE <$HISTFILE>"           Mapset <PERMANENT> in <demolocation>
HISTSIZE <3000> SAVEHIST <0>, HISTFILE </home/.../PERMANENT/.zsh_history>
```

nilason commented:
> SAVEHIST is 0. Changing it to non-zero fixes the issue for me. Outside of the GRASS session, in zsh, it is 1000.

Thanks for checking! Suspected something like that. Shouldn't be too complicated to fix. Will put a new PR.

neteler commented:
Shall this be backported, @nilason ?

nilason commented:
That would be preferable :), zsh is the default mac terminal shell from latest OS version (10.15). This might have side effect on bash too, but is on master for two weeks now, so maybe safe. In any case, a backport label could do for now.

neteler commented:
I have backported it, please test it.

nilason commented:
> I have backported it, please test it.

Great, thanks! Will test it.
